Pazniak an11 16 homework

Y-Pazniak opened the pull request:
Отправляю для проверки задание по 16 лекции.

Y-Pazniak committed:
Homework 15 (without task5)

Y-Pazniak committed:
Добавил 5-й класс.
Поправил наследование (добавил функции, доработал поля классов).
Добавил возможность добавлять случайный напиток (раньше только виски с колой).
Добавил возможность удалять напитки.
Поправил вывод напитков (добавил вывод их функций в текст вью).
Поправил функции-расширения.
Внес несущественные правки в мейн активити.

Y-Pazniak committed:
Исправлен баг с первым нажатием на кнпоку "previous", когда приходилось дублировать нажатие

Y-Pazniak committed:
ДЗ 16 (без 5 задачи).
Задачу 5 (наблюдатель паттерн) я не сделал пока - постараюсь завтра.
Остальные задачи выполнил.
1. Полям классов из предметной области переопределить геттеры и сеттеры. Сделал: добавил сеттер с проверкой корректности цифр и геттер с увеличением цены на процент НДС.
2. Создать класс каталог содержащий коллекцию объектов. Класс каталог должен быть Generic.
Сделал. Класс CatalogClass.kt
3. Добавить в main activity текстовое поле в которое необходимо выводить содержимое каталога. Сделал: вьюшка textViewCatalogContent
4а. Кнопка add должна брать первый элемент из каталога и добавлять его в singleton.
Сделал: функция addNewDrink. Отошел от задания, добавил удаление первого элемента из класса CatalogClass.kt
4б. Убрать кнопку next и previous.
Сделал. Можно еще delete убрать - все равно больше не работает.
4с. В текстовое поле из ДЗ 15 выводить весь список из singleton.
Сделал: функция printAllInfoFromSingleton

Y-Pazniak committed:
Добавлен observer (работает некорректно! доработать!)

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/homework15/data/Beer.kt:
```diff
@@ -0,0 +1,4 @@
+package by.evgen.an11pazniakhomework14.homework15.data
+
+class Beer {
```
you can remove {} if body class empty

Y-Pazniak replied:
Done, thanks for advice.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/homework15/data/Vodka.kt:
```diff
@@ -0,0 +1,31 @@
+        get() {
+            return (field * tax).toInt()
+        }
+
```
Add only one empty string after close body function

Y-Pazniak replied:
Done, thank you.

Y-Pazniak committed:
ДЗ 16 c observer
5. Настроил работу observer.
В textViewDrinkResult выводится название напитка, добавленного в синглтон сразу после нажатия на кнопку add.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/homework15/util/Util.kt:
```diff
@@ -0,0 +1,50 @@
+import by.evgen.an11pazniakhomework14.old.InheritCatNoDataClass
+
+fun WhiskeyCola.getInfo() =
+    "name: " + this.name + ", capacity: " + this.volume + ", alcohol: " + this.alco
```
you can use "name:  ${this.name}, capacity: ${this.volume }, alcohol:  ${this.alco}"

Y-Pazniak replied:
Done, thanks.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/view/MainActivity.kt:
```diff
@@ -0,0 +1,286 @@
+    private lateinit var catalogTextView: TextView
+    private var iter = 0
+    private var catalogClass: CatalogClass<Spirits> = CatalogClass()
+//    private var isFirstClick = true
```
remove comments

Y-Pazniak replied:
Comments in task-16 was necessary code for 14 and 15 tasks. Functionality of 14 and 15 tasks not used at task 16 (example - add buttons in task 15, remove buttons in task 16). One of the condition of the tasks was to improve our app, not to write new. That is why I did not remove this code, only commented before commit and push. 
Now i fixed it and removed all comments. Thank you.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/view/MainActivity.kt:
```diff
@@ -0,0 +1,286 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        findFibCycle(20) //поиск Фибоначчи в цикле
```
20 to const

Y-Pazniak replied:
Done: moved to the companion object.  Thank you.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/view/MainActivity.kt:
```diff
@@ -0,0 +1,286 @@
+        fibonacciRes.text = result
+    }
+
+    private fun recursion(countFib: Int): Int {
```
something like:
private fun recursion(countFib: Int) = when (countFib) {
             0 -> 0
             1 -> 1
             else -> (recursion(countFib - 1) + recursion(countFib - 2))
         }

Y-Pazniak replied:
Done. Thank you for advise.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/view/MainActivity.kt:
```diff
@@ -0,0 +1,286 @@
+                result = spirit.name + "\n"
+            }
+        }
+        return result
```
return result + "\n"

Y-Pazniak replied:
Shameful mistake. Fixed. Thank you.

violettazamana commented on app/src/main/java/by/evgen/an11pazniakhomework14/view/MainActivity.kt:
```diff
@@ -0,0 +1,286 @@
+
+    private fun getInfoAboutDrink(spiritArg: Spirits): String {
+        var result = ""
+        when (val spirit: Spirits = spiritArg) {
```
result = when (val spirit: Spirits = spiritArg) {

Y-Pazniak replied:
Simplified by little other way. Thank you.

Y-Pazniak committed:
"you can remove {} if body class empty"
Done.

"Add only one empty string after close body function"
Done.

"you can use "name: ${this.name}, capacity: ${this.volume }, alcohol: ${this.alco}"
Done.

"remove comments"
Done.

"20 to const"
Done.

"something like:" (rewrite recursion function shortly)
Done.

"return result + "\n"
Done.

"result = when (val spirit: Spirits = spiritArg) {"
Done.
